compiler/src/main/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGenerator.kt

```diff
@@ -0,0 +1,492 @@
+          ?: return@mapNotNull null
+
+        val modules = annotation
+          .findAnnotationArgument<KtCollectionLiteralExpression>("modules", 2)
+          ?.toFqNames(module)
+          ?: emptyList()
+
+        val exclude = annotation
+          .findAnnotationArgument<KtCollectionLiteralExpression>("exclude", 3)
+          ?.toFqNames(module)
+          ?: emptyList()
+
+        Contribution(
+          clazz = requireNotNull(clazz.getClassId()),
+          scope = clazz.scope(contributesSubcomponentFqName, module),
+          parentScope = clazz.parentScope(contributesSubcomponentFqName, module),
+          modules = modules,
+          exclude = exclude
+        )
+      }
+
+    val newContributions = contributions
+      .filterNot { contribution ->
+        val allMatchingTriggers = triggers.filter { it.scope == contribution.parentScope }
+        val excludedInAllTriggers = allMatchingTriggers.all { trigger ->
+          contribution.clazzFqName in trigger.exclusions
+        }
+
+        // If the contributed subcomponent is excluded in all triggers, then we don't want to
+        // generate any code for it and not add it to our newContributions list.
+        excludedInAllTriggers
+      }
+
+    contributions -= newContributions.toSet()
+
+    return newContributions
+      .map { contribution ->
+        val generatedPackage = contribution.generatedPackage
+        val componentClassName = contribution.generatedSubcomponentClassName
+        val factoryClass = findFactoryClass(contribution, module)
+
+        val content = FileSpec.buildFile(generatedPackage, componentClassName) {
+          TypeSpec
+            .interfaceBuilder(componentClassName)
+            .addSuperinterface(contribution.clazz.asClassName())
+            .addAnnotation(
+              AnnotationSpec
+                .builder(MergeSubcomponent::class)
+                .addMember("scope = %T::class", contribution.scope.asClassName(module))
+                .apply {
+                  fun addClassArrayMember(
+                    name: String,
+                    fqNames: List<FqName>
+                  ) {
+                    if (fqNames.isNotEmpty()) {
+                      val classes = fqNames.map { it.asClassName(module) }
+                      val template = classes
+                        .joinToString(prefix = "[", postfix = "]") { "%T::class" }
+
+                      addMember("$name = $template", *classes.toTypedArray())
+                    }
+                  }
+
+                  addClassArrayMember("modules", contribution.modules)
+                  addClassArrayMember("exclude", contribution.exclude)
+                }
+                .build()
+            )
+            .apply {
+              if (factoryClass != null) {
+                addType(generateFactory(factoryClass.originalDescriptor))
+              }
+            }
+            .addType(generateParentComponent(contribution, module, factoryClass))
```

Comment:
Prior to this change, we had something like this in **application** code:
```kotlin
@MergeSubcomponent(FragmentScope::class)
@SingleIn(FragmentScope::class)
public interface GeneratedFragmentComponent : FragmentComponent {

    @Subcomponent.Factory
    public interface Factory : FragmentComponent.Factory

    @ContributesTo(ActivityScope::class)
    @dagger.Module(subcomponents = [GeneratedFragmentComponent::class])
    public interface Module {

        @Binds
        public fun bindComponent(target: Factory): FragmentComponent.Factory // 1
    }
}
```
which enabled us inject a `FragmentComponent.Factory` directly when provisioning a fragment. After trying out this PR, I modified the **library** code to define the relationship directly. So it became something like this
```kotlin
@ContributesSubcomponent(
    scope = FragmentScope::class,
    parentScope = ActivityScope::class
)
@SingleIn(FragmentScope::class) // 2
public interface FragmentComponent {
    public fun getFragmentMap(): FragmentMap

    @ContributesSubcomponent.Factory
    public interface Factory {
        public fun create(@BindsInstance someParam: SomeParam): FragmentComponent
    }

    @ContributesTo(ActivityScope::class)
    public interface ParentComponent {
        public fun createFragmentFactory(): Factory
    }
}
```
which generated code that looks like this:
```kotlin
@MergeSubcomponent(scope = FragmentScope::class)
public interface FragmentComponentAnvilSubcomponent : FragmentComponent {
  @Subcomponent.Factory
  public interface SubcomponentFactory : FragmentComponent.Factory

  @ContributesTo(
    ActivityScope::class,
    replaces = [FragmentComponent.ParentComponent::class]
  )
  public interface ParentComponent : FragmentComponent.ParentComponent {
    public override fun createFragmentFactory(): SubcomponentFactory
  }
}
```

A few things stand out here:
1. We had a module that was contributed into the parent scope, where we could bind the generated Factory with the defined one. This was how we were able to simply inject the defined Factory. With this PR, we couldn't do that anymore even if we wanted to, since library code doesn't have access to the generated Factory from application code. I think it will be desirable to generate this auto-wiring module here as well
2. Since ideally, we can now totally avoid writing a subcomponent, I tried adding the `SingleIn` scope annotation on the FragmentComponent, and hoped that it would also be copied over to the generated FragmentComponentAnvilSubcomponent, but that wasn't the case. I wonder if this is something that can/should happen

Reply:
I appreciate the very detailed feedback a lot 🙏🏻 

For 1) I haven't tested it yet, but I assumed that you'd be able to inject the factory. I'll double check, thanks for pointing out!

2) Is an oversight, I'll fix that.

Reply:
1) is addressed here https://github.com/square/anvil/pull/427 It was quite easy to solve.

Reply:
2) is addressed here https://github.com/square/anvil/pull/428

Reply:
Beautiful!
That was really quick. Thank you
Went through both PRs, and they look solid to me ✅ 

If you get a chance to publish a new snapshot once they get merged, that'd be nice. I can do another quick test and let you know how it goes

Reply:
There was actually another big issue that I had to fix: https://github.com/square/anvil/pull/430

I published a new snapshot ([`2.3.10-SNAPSHOT`](https://oss.sonatype.org/content/repositories/snapshots/com/squareup/anvil/compiler/2.3.10-SNAPSHOT/)) and tested the feature finally internally myself and our first use case can be solved with this new feature. More testing next week.

Please let me know how the snapshot works for you.

Reply:
Oh. Great find! Probably would have run into that eventually. Haha
I played with the new snapshot, and it worked great

I'll give it a spin on a much larger codebase and report any findings. Really looking forward to having this feature merged in 🚀